Extract trading UI to separate component

dinesh-skyach opened the pull request:
[WIP] Issue-347 Extract trading UI Refracting

dinesh-skyach commented:
Hi @yivo , I have strictly followed your steps. Redirect all request of trading UI to Private::MarketsController. Please let me know if we need move that code to JWT.
I have also added CORS to gemfile, since on local we have to handle these on different ports locally. But on production it may not be required.
Please find the trading UI code at the following repository:-
https://github.com/skyach/peatio-trading-ui

Please let me know if you still need improvements. 

@yivo , One more question I had, how are you going to link there 2 apps in development environment. For linking, I think, we ned to define the linking url in YML, which devs can change according to the domains, as on local it will behave differently than production.

Also, for the failing test cases, we either need to remove them from here or we need to add them in the trading ui app.

yivo commented:
Hello @dinesh-skyach 

Can you please remove rack-cors? It is not required since we have this already implemented in grape APIs (check API_CORS_ORIGINS). I would like to see the final code without dev stuff and then if it would be needed I will assign someone. 

About tests – I think we need to remove them since I am sure it will be more unpossible to move them to peatio-trading-ui.

dinesh-skyach commented:
Hi @yivo , I have removed the development stuff and comments. I also have removed the used tests related to the moved code.

You may face problems in linking configuration, please let me know, if I can be of help to anyone, as I am not sure how you are going to present both repos, since this is not a seemless integration as we do in case of gems generally as there is a gap between local and production.

On Tests, I did not mean to shift them as it is, but to write on the peation-trading-ui separately. But anyways, I have removed them from here.

yivo commented on app/views/shared/_header.html.erb:
```diff
@@ -4,7 +4,7 @@
         <img class="nav-img" src="<%= ENV['LOGO_URL'] || image_path('logo.svg') %>">
       </a>
-      <%= top_nav_link(t('header.market'), market_path(current_market.id), 'exchange', target: '_blank') if current_user %>
+      <%= top_nav_link(t('header.market'), market_path(current_market.id), 'exchange', target: '_blank') if current_user%>
```
@dinesh-skyach Can you please rollback this change?

dinesh-skyach replied:
@yivo done

yivo commented on app/controllers/private/orders_controller.rb:
```diff
@@ -1,6 +1,8 @@
 module Private
   class OrdersController < BaseController
 
+    skip_before_action :verify_authenticity_token
```
@dinesh-skyach Can you keep CSRF protection enabled? Try to send appropriate cookies from peatio-trading-ui and it will work.

dinesh-skyach replied:
@yivo done

yivo commented on app/controllers/private/order_bids_controller.rb:
```diff
@@ -2,6 +2,8 @@ module Private
   class OrderBidsController < BaseController
     include Concerns::OrderCreation
 
+    skip_before_action :verify_authenticity_token
```
@dinesh-skyach Can you keep CSRF protection enabled? Try to send appropriate cookies from peatio-trading-ui and it will work.

dinesh-skyach replied:
@yivo done

yivo commented on app/controllers/private/order_asks_controller.rb:
```diff
@@ -2,6 +2,8 @@ module Private
   class OrderAsksController < BaseController
     include Concerns::OrderCreation
 
+    skip_before_action :verify_authenticity_token
```
@dinesh-skyach Can you keep CSRF protection enabled? Try to send appropriate cookies from peatio-trading-ui and it will work.

dinesh-skyach replied:
@yivo done

dinesh-skyach commented on app/controllers/private/markets_controller.rb:
```diff
@@ -25,6 +25,9 @@ def show
       set_member_data if current_user
       gon.jbuilder
+      respond_to do |format|
+        format.json
```
As we cannot remove the route for this controller method. As this is being used from new Trading UI App. So now this will only respond to json requests.

dinesh-skyach commented:
Hi @yivo I have made the changes.

Please let me know if you still need improvements.

yivo commented:
@dinesh-skyach Have you migrated images & locales? Can you get access tokens at test.bitgo.com and test deposits / withdraws / trading & admin?

dinesh-skyach commented:
@yivo I have migrated images and locales. I have tested the site via adding balance from backend but will test the site thoroughly via using access tokens from test.bitgo.com.

yivo commented:
@dinesh-skyach Do you need any help with bitgo?

dinesh-skyach commented:
@yivo I have tested the site and everything is working accordingly.

dinesh-skyach commented:
@yivo I have already rebased both trading UI and PR?

dinesh-skyach commented:
Hi @yivo , as per today’s call, are we going to make this a gem instead of a separate application? Or we are keeping this as it is? 
Please let me know if we need to change this.

yivo commented:
@dinesh-skyach Hello. We are going to extract UI to Rails Engine but keep trading as separate Rails app.

yivo commented:
So we are moving to the same direction.

yssavchuk commented:
@dinesh-skyach Hi!
I have run your https://github.com/skyach/peatio-trading-ui and rapidly took a look
Here is few drawback I found

- [ ] BTC html etc
![deepinscreenshot_select-area_20180208172159](https://user-images.githubusercontent.com/19859122/35981744-b422398a-0cf6-11e8-9da6-aa42fc728d8c.png)

- [ ] Buy USD for USD
![deepinscreenshot_select-area_20180208172249](https://user-images.githubusercontent.com/19859122/35981819-e3deeca4-0cf6-11e8-94cd-53fc6b1fcaf2.png)

- [ ]  Looks like this button changes after click (compare with original peatio trading)
![deepinscreenshot_select-area_20180208172655](https://user-images.githubusercontent.com/19859122/35981863-092dac7a-0cf7-11e8-8805-51fd0d7191fb.png)

- [ ] Dash image was lost and LTC looks badly
![deepinscreenshot_select-area_20180208172740](https://user-images.githubusercontent.com/19859122/35981908-229871ae-0cf7-11e8-90c1-47fd5238641f.png)

- [ ] USD markets filter doesn't work.  I have already fixed this in https://github.com/rubykube/peatio/pull/472 take a look

dinesh-skyach commented:
Hi @yssavchuk, Thanks for your feedback. I have fixed all the above issues. Kindly have a look and let me know if you have any feedback.

@yssavchuk  Also one question for the second issue "Buy USD for USD" I didn't get this. As we are buying BTC from USD, So how can this be BUY USD.

yivo commented:
@dinesh-skyach Can you please rebase? We have now changes.

dinesh-skyach commented:
Hi @yivo I have already rebased.

yssavchuk commented:
@dinesh-skyach In the second issue I mean that price shoul be in USD and amount is BTC 
![deepinscreenshot_select-area_20180208172249](https://user-images.githubusercontent.com/19859122/36037738-5a402852-0dc6-11e8-8152-479e81953d24.png)

yivo commented:
@dinesh-skyach @yssavchuk Also these labels were uppercased before.

dinesh-skyach commented:
@yivo @yssavchuk I have already fixed this in above commits.

yivo commented on lib/generators/config/templates/application.yml.erb:
```diff
@@ -128,6 +125,9 @@ development:
   AUTH0_OAUTH2_CLIENT_SECRET: LxUPnrrrpRHSB-vVgWU3TwmakQ8acY5mYVTaVgkREQL3i9avjpOnWTXSSEsrbLzK
 
+  # Configure markets page url:
+  MARKETS_HOST: # Marketing URL of peatio (eg. "http://localhost:3001")
```
@dinesh-skyach Host != URL. Please update.

dinesh-skyach replied:
@yivo I have made the change

yivo commented on app/controllers/application_controller.rb:
```diff
@@ -179,4 +180,15 @@ def clear_all_sessions(member_id)
+    response.headers.merge!(cors_headers) if Rails.env.development?
+  end
+
+  def cors_headers
```
@dinesh-skyach Please, use single quotes. Move this to markets controller.

dinesh-skyach replied:
@yivo As this method is being called from multiple controllers ('order_asks controller', 'order_bids controller', 'orders controller') and is not being used in markets controller. So this is the only reason I have added this to application controller.

dinesh-skyach replied:
@yivo Any thoughts on this?

yivo commented on app/controllers/private/markets_controller.rb:
```diff
@@ -21,9 +21,13 @@ def show
       # default to limit order
       @order_bid = OrderBid.new ord_type: 'limit'
       @order_ask = OrderAsk.new ord_type: 'limit'
+      @markets_data = @markets.map{|m| {market: {name: m.name, id: m.id, quote_unit: m.quote_unit.upcase == Peatio.base_fiat_ccy ? 'fiat' : m.quote_unit}, global_currency: Global[m.id].ticker[:last]} }
```
@dinesh-skyach Pls move it to method.

dinesh-skyach replied:
@yivo I have made the change

yivo commented on app/helpers/application_helper.rb:
```diff
@@ -253,4 +253,8 @@ def root_url_with_port_from_request
     parts.join('')
   end
+
+  def set_market_path
```
@dinesh-skyach Can you cleanup this? market_path may return URL actually depending on configuration. Also pls reduce the complexity.

dinesh-skyach replied:
@yivo I have made the changes

dinesh-skyach commented:
@yivo I have made the changes. 

Let me know if still improvements are required.

dinesh-skyach commented:
Hi @yivo, I have rebased the PR. Also there are some changes to be made in petio trading UI. I will do this tomorrow.

dinesh-skyach commented:
@yivo I have rebased the PR and Peatio Trading UI.

mod commented:
@dinesh-skyach @yivo This task becomes urgent, what is our ETA and challenges to get this merged?
Do you guys think we can turn this into a rails engine installable on target rails?

yivo commented:
@mod It sound logical to have tests in UI repository but it is not possible right since Peatio specs use lot of stubbing for these specs:
![image](https://user-images.githubusercontent.com/7421323/37458994-6cf7ab8e-284f-11e8-9d10-dd048500ec16.png)
<img width="690" alt="2018-03-15--12-49-37" src="https://user-images.githubusercontent.com/7421323/37459001-72d822ae-284f-11e8-9f63-ed7da8402f2c.png">
![image](https://user-images.githubusercontent.com/7421323/37459015-80b48070-284f-11e8-917a-e72f4767acc4.png)

mod commented:
@yivo we need all the trading tests to be made using faraday can you open an issue ?
